Wrappers/Python/cil/optimisation/utilities/preconditioner.py

```diff
@@ -0,0 +1,179 @@
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Authors:
+# - CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+
+
+from abc import ABC, abstractmethod
+import numpy as np
+
+
+class Preconditioner(ABC):
+
+    r"""
+    Abstract base class for Preconditioner objects. The `apply` method of this class takes an initialised CIL function as an argument and modifies a provided  `gradient`.
+   
+    
+    Methods
+    -------
+    apply(x)
+        Abstract method to call the preconditioner.
+    """
+
+    def __init__(self):
+        '''Initialises the preconditioner
+        '''
+        pass
+
+    @abstractmethod
+    def apply(self, algorithm, gradient, out=None):
+        r"""
+        Abstract method to apply the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : Algorithm
+            The algorithm object.
+    
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, optional, default is None 
+            The modified gradient to return 
+            
+        """
+        pass
+
+
+class Sensitivity(Preconditioner):
+
+    r"""
+    Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` y/(A^T \mathbf{1})` where :math:`A` is an operator, :math:`\mathbf{1}` is an object in the range of the operator filled with ones and :math:`y` is an optional reference image.  
+
+    Parameters
+    ----------
+    operator : CIL Operator 
+        The operator used for sensitivity computation.
+    reference : DataContainer e.g. ImageData, optional
+        The reference data, an object in the domain of the operator. Recommended to be a best guess reconstruction. 
+    """
+
+    def __init__(self, operator, reference=None):
+
+        super(Sensitivity, self).__init__()
+        self.operator = operator
+        self.reference = reference
+
+        self.compute_preconditioner_matrix()
+
+    def compute_preconditioner_matrix(self):
+        r"""
+        Compute the sensitivity. :math:`A^T \mathbf{1}` where :math:`A` is the operator and :math:`\mathbf{1}` is an object in the range of the operator filled with ones.        
+        Then perform safe division by the sensitivity to store the preconditioner array :math:` y/(A^T \mathbf{1})`
+        """
+        self.array = self.operator.adjoint(
+            self.operator.range_geometry().allocate(value=1.0))
+
+        if self.reference is not None:
```

Comment:
Do we want this "safe division" where we only divide if non-zero and leave ones if not, or do we want to add a small epsilon value?